Reader: Add new header style for the tag stream page

dvdchr opened the pull request:
Refs pcdRpT-3Eb-p2
Depends on #21475

This partially implements the new header style for the Tag Stream page. I tried using the SwiftUI but it's not going to work without a decent refactor in the `ReaderStreamViewController+Helper` because it expects the header to be a `UIView` that conforms to `ReaderStreamHeader`.

Additionally, I tried implementing the horizontal layout by following a simple rule: we'll use the horizontal layout if the tag title is less than 10 characters AND the user doesn't use any accessibility categories. Let me know how it works, I can revert it and always use vertical layout if it's clunky. cc: @osullivanchris 

Here are some previews:

### Before

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/28991e7e-2dcf-42ed-b6c3-90f38d9dcd59" width=350>

### After

Short titles | Longer titles
-|-
![Simulator Screenshot - iPhone 14 Pro - 2023-08-31 at 18 30 00](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/952ebc16-44a4-4c06-a48e-db2375c77684) | ![Simulator Screenshot - iPhone 14 Pro - 2023-08-31 at 18 30 47](https://github.com/wordpress-mobile/WordPress-iOS/assets/1299411/5bca8121-d0b6-4ca1-ae2f-d944b22ea3c1)

⚠️ A couple notes ⚠️ 

- The Follow/Unfollow button is still using the old design, as per the breakdown suggested in p1693342310830629/1693330435.413469-slack-C05N140C8H5.
- The metadata information is not included yet in this PR, as it requires an additional API endpoint.
- I noticed that the header has a bottom border on the design. However, I decided to exclude this for now because we are still using the "old" layout for the overall stream page; there is a gap between the header and each card. Let me know if you'd prefer the border to be added!

## To test

- Launch the Jetpack app.
- Make sure the `Reader Improvements v1` feature flag is turned on.
- Go to Reader > Discover, and select any recommended tag.
- 🔎 Verify that the new header design is shown.
- 🔎 Verify that the follow/unfollow functionality works.
- Go back and try finding a tag with longer text. Usually, you can find these displayed on the top section of any reader card.
- 🔎 Verify that header section is displayed using vertical layout.

## Regression Notes
1. Potential unintended areas of impact
Should be none. The feature is hidden behind a feature flag.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes.

3. What automated tests I added (or what prevented me from doing so)
N/A.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dvdchr committed:
Show the new header style when the feature flag is enabled

osullivanchris reviewed (approved):
Looks good. Yep the button can follow, and the border is relevant to the new card stream design, but not needed for now.

One small thing - can you remove the title 'Topic' from the navigation bar? It is not useful/necessary.

wargcm commented on WordPress/Classes/ViewRelated/Reader/ReaderTagFeedHeader.swift:
It looks like this file was committed accidentally.

dvdchr replied:
Ah, thank you for noticing. I'll revert this!

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/21477/commits/52cb1865c5f275739b2fba8521c19bdab2875bf5.

wargcm commented on WordPress/WordPress.xcodeproj/project.pbxproj:
Should the Cocoapods changes be here? They don't appear to be causing any build issues since your builds passed, so I'll leave it up to you whether you want to keep them or not.

dvdchr replied:
I'll try to revert these. Thanks for noticing!

dvdchr replied:
Done in https://github.com/wordpress-mobile/WordPress-iOS/pull/21477/commits/3ad15f44d2528455691bbdf9dae4151870ff59f4.

wargcm reviewed (approved):
I noticed that there are breaking constraints with this header:
<details><summary>Constraints</summary>

```
[LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x136ed7330 UILabel:0x136ea2d10.leading == WordPress.ReaderTagStreamHeader:0x136efb630.leadingMargin   (active)>",
    "<NSLayoutConstraint:0x136e261c0 H:[UILabel:0x136ea2d10]-(>=8)-[UIButton:0x135edcb30'Follow']   (active)>",
    "<NSLayoutConstraint:0x136ed7230 WordPress.ReaderTagStreamHeader:0x136efb630.trailingMargin == UIButton:0x135edcb30'Follow'.trailing   (active)>",
    "<NSLayoutConstraint:0x135eb5730 'UIView-Encapsulated-Layout-Width' WordPress.ReaderTagStreamHeader:0x136efb630.width == 0   (active)>",
    "<NSLayoutConstraint:0x136efa2a0 'UIView-leftMargin-guide-constraint' H:|-(8)-[UILayoutGuide:0x135e42c70'UIViewLayoutMarginsGuide'](LTR)   (active, names: '|':WordPress.ReaderTagStreamHeader:0x136efb630 )>",
    "<NSLayoutConstraint:0x136ecbb60 'UIView-rightMargin-guide-constraint' H:[UILayoutGuide:0x135e42c70'UIViewLayoutMarginsGuide']-(8)-|(LTR)   (active, names: '|':WordPress.ReaderTagStreamHeader:0x136efb630 )>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x136e261c0 H:[UILabel:0x136ea2d10]-(>=8)-[UIButton:0x135edcb30'Follow']   (active)>

----

[LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x136ec8a80 V:|-(16)-[UILabel:0x136ea2d10]   (active, names: '|':WordPress.ReaderTagStreamHeader:0x136efb630 )>",
    "<NSLayoutConstraint:0x136e8bf60 V:[UILabel:0x136ea2d10]-(16)-|   (active, names: '|':WordPress.ReaderTagStreamHeader:0x136efb630 )>",
    "<NSLayoutConstraint:0x135ebd920 'UIView-Encapsulated-Layout-Height' WordPress.ReaderTagStreamHeader:0x136efb630.height == 0   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x136e8bf60 V:[UILabel:0x136ea2d10]-(16)-|   (active, names: '|':WordPress.ReaderTagStreamHeader:0x136efb630 )>
```

</details>

The new header appears correct, and it looks like this was already happening for the old header too. Given that, I don't think it's a must fix and more of a nice to have.

The rest LGTM 👍

dvdchr committed:
Remove horizontal layout configuration

dvdchr committed:
Titleize the tag title

dvdchr committed:
Remove navigation title from the tag stream

dvdchr committed:
Remove unused file

dvdchr committed:
Revert pbxproj changes

dvdchr commented:
Some additional changes:

- https://github.com/wordpress-mobile/WordPress-iOS/pull/21477/commits/65112985a2d7cb55ae15cc24a0062a5eb3b19e7d The horizontal layout configuration is now removed. The tag stream header will always use the vertical layout for now.
- https://github.com/wordpress-mobile/WordPress-iOS/pull/21477/commits/1a6384969ea8097a53f65263dd84c2981dbbb6e1 Tag stream titles are now titleized properly., e.g. `great-outdoors` -> `Great Outdoors`
- https://github.com/wordpress-mobile/WordPress-iOS/pull/21477/commits/0a4c60b93b496c09557c80ba7e6602f00bca8ab2 "Topic" navigation title is now removed in the tag stream.

FYI: I am enabling auto-merge.
